Expand relative dev paths in environment files

haampie opened the pull request:
Solves an issue where `spack env create [name] [path]` followed by `spack -e [name] install` fails when there are relative paths in `develop:spec:path`.

The problem is caused by spack relocating the spack.yaml file to `<spack>/var/environments/<name>/spack.yaml`.

This PR expands the relative paths to their normalized absolute path after reading the spack.yaml file.

Also fixes an issue for anonymous environments where relative paths ended up properly expanded in the `dev_path` variant of specs, but they were not normalized, leading to variants like `myspec dev_path=/path/to/env/dir/../sources`.

haampie committed:
Fix redundant isabs and always normalize.

* joinpath handles absolute paths of 2nd arg.
* avoid ending up with things like
  dev_path=/some/path/to/env/../build
  which is absolute but not normalized

haampie committed:
Rewrite relative dev_spec paths internally to absolute paths in case of relocation of the environment file

haampie committed:
Test relative paths for dev_path in environments

haampie committed:
Test failed for me because the export PATH=... bit is quoted

haampie committed:
Fix typo

haampie committed:
Update the dev_build test to use relative paths (fails without parent commits)

haampie committed:
make flake happy

haampie committed:
Make pyton os.path.relpath happy

haampie committed:
Fix tests on some platforms

haampie committed:
Fix comment

becker33 commented:
This will cause problems for anyone trying to share an environment. The use case for relative paths is that I can take the environment and give it to you, and you can run `spack develop` to download the dev packages and then `spack install` to install them. So I think this will require more thought about how the interface should work.

haampie commented:
Hm yeah, is it really necessary to relocate these spack.{yaml,lock} files when running `spack env create`?

I rarely use `spack env create`, it was just that a test was failing when I used `path: ../build` as a dev path.

haampie committed:
refactor and add tests

haampie commented:
@becker33: anonymous environments will never rewrite relative to absolute paths since they are never constructed with an `init_file`. So only when promoting an anonymous environment to a named environment or when using the `spack env create -d ...` flag this rewrite will kick in. Just to be sure I'm only rewriting paths when `env.path` is different from `spack.yaml`'s directory.

See the tests for when paths are rewritten and when they are not.

So, I think this PR is done and it fixes effectively two things:
- Rewrite relative develop paths to absolute paths when initializing a new environment from a spack.yaml in a different directory.
- Normalize `dev_path`, e.g. `spec dev_path=/path/to/xyz/../build` -> `spec dev_path=/path/to/build`

Both changes are tested.

haampie committed:
Fix python 2.7

becker33 commented on lib/spack/spack/environment.py:
```diff
@@ -621,6 +621,10 @@ def __init__(self, path, init_file=None, with_view=None):
                 else:
                     self._read_manifest(f, raw_yaml=default_manifest_yaml)
+
+                if hasattr(f, 'name') and f.name.endswith('.yaml'):
+                    init_file_dir = os.path.abspath(os.path.dirname(f.name))
+                    self._rewrite_relative_paths_on_relocation(init_file_dir)
```
This needs a comment to explain why it's checking the file for a name.

Additionally, there should probably be an option to determine whether this is done. I'm picturing two workflows.

1. I create an environment and use it for a while, it points to repos where I have work and I want to keep working in those repos, but now I want the environment to be managed by Spack so that it can use reference counting to ensure I don't uninstall anything I need. I want the paths to be rewritten when I run `spack env create foo /path/to/spack.yaml`
2. I'm given an environment, and I want my copy of it to be managed by Spack. So the first thing I do before downloading any source code is run `spack env create foo /path/to/spack.yaml`. I don't want the paths to be rewritten in this case, the next thing I'm going to do is tell the new environment to download its own copies.

haampie replied:
Right, another thing is you'd probably would not want to run `spack create my_env ./spack.yaml` when `spack.yaml` has a relative develop path that escapes the environment directory `../xyz`.

I just defaulted to expanding the develop path and added `--keep-relative` to force spack to copy paths verbatim.

The reason is I like to store a `spack.yaml` env file inside my git repo for either CI or development with e.g. `path: ../..`

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -2369,3 +2373,51 @@ def _write_helper_raise(self, x, y):
+    spack_yaml = init_env.join('spack.yaml')
+    spack_yaml.write(raw_yaml)
+
+    # Create new environment in different directory; should rewrite rel paths
+    env('create', '-d', str(dest_env), str(spack_yaml))
+    with ev.Environment(str(dest_env)) as e:
+        assert e.dev_specs['mypkg1']['path'] == str(build_folder)
+        assert e.dev_specs['mypkg2']['path'] == '/some/other/path'
+
+    # Create new environment in var/share/environments; should rewrite rel paths
+    env('create', 'named_env', str(spack_yaml))
+    with ev.read('named_env') as e:
+        assert e.dev_specs['mypkg1']['path'] == str(build_folder)
+        assert e.dev_specs['mypkg2']['path'] == '/some/other/path'
+
+    # "Create" environment in original directory; should not rewrite rel paths
+    with ev.Environment(str(init_env), str(spack_yaml)) as e:
+        assert e.dev_specs['mypkg1']['path'] == '../build_folder'
+        assert e.dev_specs['mypkg2']['path'] == '/some/other/path'
+
+    env('create', '-d', str(init_env), str(spack_yaml))
+    with ev.Environment(str(init_env)) as e:
+        assert e.dev_specs['mypkg1']['path'] == '../build_folder'
+        assert e.dev_specs['mypkg2']['path'] == '/some/other/path'
```
It will be easier to diagnose what is broken if these are separate tests

haampie committed:
Break up tests

haampie committed:
Add a --keep-relative flag to spack env create

This ensures that relative paths of develop paths are not expanded to
absolute paths when initializing the environment in a different location
from the spack.yaml init file.

becker33 commented on lib/spack/spack/concretize.py:
```diff
@@ -73,9 +73,7 @@ def concretize_develop(self, spec):
-        path = dev_info['path']
-        path = path if os.path.isabs(path) else os.path.join(
-            env.path, path)
+        path = os.path.normpath(os.path.join(env.path, dev_info['path']))
```
Is this relying on the fact that `os.path.join` returns the second argument if it is absolute?

haampie replied:
Yeah, should also work on Windows iiuc: https://docs.python.org/3/library/os.path.html#os.path.join

>  If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component.

haampie commented:
Thanks! :)
